### PR TITLE
DAO

### PR DESCRIPTION
**Pull request by Theo-Hafsaoui**
## DAO
To begin complete the user story we need the DAO classes
image-title
**Issue:** #6 
image-title
**Issue:** #8

### CODE REVIEW

**Comment by anastygnome** [comment]
@Theo-Hafsaoui can you review ?

**Comment by Theo-Hafsaoui** [comment]
not yet i will make it a reviewable when it's ready

**Comment by Theo-Hafsaoui** [comment]
we are **far** from over on this PR

**Comment by Theo-Hafsaoui** [comment]
Stop, pushing for review it's not done yet

**Comment by anastygnome** [comment]
I'm asking you to review the last commit, you can do a per-commit review.

**Comment by Theo-Hafsaoui** [comment]
I see... my apologies. i already the test this morning before the amend it worked.
i hope the new one too

File: edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SubjectDAO.java [hunk 1]
```diff
@@ -25,7 +26,7 @@ public class SubjectDAO implements DAO<Subject> {
      * Constructor of SubjectDao, initiate connection and prepared statement
      * @throws SQLException throw a SQLException if there is a problem with the connection or database prepared statement
      */
-    private SubjectDAO() throws SQLException {
+    public SubjectDAO() throws SQLException {
```
Comment (anastygnome) [inline edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/slots/SubjectDAO.java:29]: This  basically negates the whole idea of a connection pool, as the connection you will be using is gonna be on hold even when no operation is pending (its lifetime follows that of the class)

What you should instead do is 
1. make each of those static and or final strings 
2.  Make each of the method throw an Exception (whatever really)
3. in each method, use a try with ressources to get the conn. and the prepare statement :  the following works : 
```java 
//this.<whatever> := "Select from Subject where id = (?) "
//...
try( 
      Connection connection = Datasource.getConnection(); // I've added this in the Datasource in an early PR
       PreparedStatement stmt = connection.prepareStatement(this.<whatever>) // no ; required on the last one
       ) 
       {  /*resultSet.Set(...) or Get*/ }
catch (SqlException e) { throw e; /* or whatever */}

I see this is applicable to other DAOs as well, I can provide help if you so wish.

File: edtLHD/src/test/java/fr/univtln/lhd/model/entities/dao/slots/SubjectDAOTest.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+        try {
+            st = new SubjectDAO();
+        } catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+        return st;
+    }
+
+
+    private Subject getSubject(){//no cons on unicity random is useless
```
Comment (anastygnome) [inline edtLHD/src/test/java/fr/univtln/lhd/model/entities/dao/slots/SubjectDAOTest.java:25]: Typo  in comment unicity  -> uniqueness 

Also, I will add such a constraint in a PR soon.

**Review by anastygnome — COMMENTED** [review]
Nice work overall, just wanna mention a few issues ahead of merge.

File: edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/user/AdminDAO.java [hunk 3]
```diff
@@ -26,18 +26,18 @@ public class AdminDAO implements DAO<Admin> {
...
-        this.getAll = connection.prepareStatement("SELECT * FROM MANAGERS");
-        this.save = connection.prepareStatement("INSERT INTO MANAGERS VALUES (DEFAULT, ?, ?, ?, ?, ?)",RETURN_GENERATED_KEYS);
-        this.update = connection.prepareStatement("UPDATE MANAGERS SET name=?, fname=? ,email=?,dpt=? WHERE ID=?");
-        this.delete = connection.prepareStatement("DELETE FROM MANAGERS WHERE ID=?");
+        this.get = connection.prepareStatement("SELECT * FROM ADMINS WHERE ID=?");
+        this.getAll = connection.prepareStatement("SELECT * FROM ADMINS");
+        this.save = connection.prepareStatement("INSERT INTO ADMINS VALUES (DEFAULT, ?, ?, ?, ?, ?)",RETURN_GENERATED_KEYS);
+        this.update = connection.prepareStatement("UPDATE ADMINS SET name=?, fname=? ,email=?,title=? WHERE ID=?");
+        this.delete = connection.prepareStatement("DELETE FROM ADMINS WHERE ID=?");
     }
```
Comment (anastygnome) [inline edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/user/AdminDAO.java:34]: This  basically negates the whole idea of a connection pool, as the connection you will be using is gonna be on hold even when no operation is pending (its lifetime follows that of the class)

What you should instead do is 
1. make each of those static and or final strings 
2.  Make each of the method throw an Exception (whatever really)
3. in each method, use a try with ressources to get the conn. and the prepare statement :  the following works : 
```java 
//this.<whatever> := "Select from Subject where id = (?) "
//...
try( 
      Connection connection = Datasource.getConnection(); // I've added this in the Datasource in an earlier PR
       PreparedStatement stmt = connection.prepareStatement(this.<whatever>) // no ; required on the last one
       ) 
       {  /*resultSet.Set(...) or Get*/ }
catch (SqlException e) { throw e; /* or whatever */}

I see this is applicable to other DAOs as well, I can provide help if you so wish.

Comment (Theo-Hafsaoui) [reply edtLHD/src/main/java/fr/univtln/lhd/model/entities/dao/user/AdminDAO.java:34]: hmmm, might do it later

**Pull request merged**